beacon-chain/rpc/beaconv1/state.go

```diff
@@ -66,41 +116,73 @@ func (bs *Server) GetStateRoot(ctx context.Context, req *ethpb.StateRequest) (*e
 	case "justified":
 		root, err = bs.justifiedStateRoot(ctx)
 	default:
-		ok, matchErr := bytesutil.IsBytes32Hex(req.StateId)
+		ok, matchErr := bytesutil.IsBytes32Hex(stateId)
 		if matchErr != nil {
-			return nil, status.Errorf(codes.Internal, "Failed to parse ID: %v", err)
+			return nil, status.Errorf(codes.Internal, "Could not parse ID: %v", err)
 		}
 		if ok {
-			root, err = bs.stateRootByHex(ctx, req.StateId)
+			root, err = bs.stateRootByHex(ctx, stateId)
 		} else {
-			slot, parseErr := strconv.ParseUint(stateIdString, 10, 64)
+			slotNumber, parseErr := strconv.ParseUint(stateIdString, 10, 64)
 			if parseErr != nil {
 				// ID format does not match any valid options.
 				return nil, status.Errorf(codes.Internal, "Invalid state ID: "+stateIdString)
 			}
-			root, err = bs.stateRootBySlot(ctx, types.Slot(slot))
+			root, err = bs.stateRootBySlot(ctx, types.Slot(slotNumber))
 		}
 	}
 
-	if err != nil {
-		return nil, err
-	}
-	return &ethpb.StateRootResponse{
-		Data: &ethpb.StateRootResponse_StateRoot{
-			StateRoot: root,
-		},
-	}, nil
+	return root, err
 }
 
-// GetStateFork returns Fork object for state with given 'stateId'.
-func (bs *Server) GetStateFork(ctx context.Context, req *ethpb.StateRequest) (*ethpb.StateForkResponse, error) {
-	return nil, errors.New("unimplemented")
-}
+func (bs *Server) state(ctx context.Context, stateId []byte) (*statetrie.BeaconState, error) {
+	var (
+		s   *statetrie.BeaconState
+		err error
+	)
 
-// GetFinalityCheckpoints returns finality checkpoints for state with given 'stateId'. In case finality is
-// not yet achieved, checkpoint should return epoch 0 and ZERO_HASH as root.
-func (bs *Server) GetFinalityCheckpoints(ctx context.Context, req *ethpb.StateRequest) (*ethpb.StateFinalityCheckpointResponse, error) {
-	return nil, errors.New("unimplemented")
+	stateIdString := strings.ToLower(string(stateId))
+	switch stateIdString {
+	case "head":
+		s, err = bs.ChainInfoFetcher.HeadState(ctx)
+		if err != nil {
+			return nil, status.Errorf(codes.Internal, "Could not get head state: %v", err)
+		}
+	case "genesis":
+		s, err = bs.StateGenService.StateBySlot(ctx, 0)
```

Comment:
Is this OK to get the genesis state or is there a better way?

Reply:
how about just directly retrieving the genesis state from the db ?